akka-stream/src/main/scala/akka/stream/scaladsl/RetryFlow.scala

```diff
@@ -0,0 +1,157 @@
+   *
+   * The wrapped `flow` and `decideRetry` take the additional context parameters which can be a context,
+   * or used to control retrying with other information.
+   *
+   * @param minBackoff minimum duration to backoff between issuing retries
+   * @param maxBackoff maximum duration to backoff between issuing retries
+   * @param randomFactor adds jitter to the retry delay. Use 0 for no jitter
+   * @param maxRetries total number of allowed retries, when reached the last result will be emitted
+   *                   even if unsuccessful
+   * @param flow a flow with context to retry elements from
+   * @param decideRetry retry condition decision function
+   */
+  @ApiMayChange
+  def withBackoffAndContext[In, CtxIn, Out, CtxOut, Mat](
+      minBackoff: FiniteDuration,
+      maxBackoff: FiniteDuration,
+      randomFactor: Double,
+      maxRetries: Int,
+      flow: FlowWithContext[In, CtxIn, Out, CtxOut, Mat])(
+      decideRetry: ((In, CtxIn), (Out, CtxOut)) => Option[(In, CtxIn)]): FlowWithContext[In, CtxIn, Out, CtxOut, Mat] =
+    FlowWithContext.fromTuples {
+      val retryCoordination = BidiFlow.fromGraph(
+        new RetryFlowCoordinator[In, CtxIn, Out, CtxOut](minBackoff, maxBackoff, randomFactor, maxRetries, decideRetry))
+
+      retryCoordination.joinMat(flow)(Keep.right)
+    }
+}
+
+/**
+ * INTERNAL API.
+ */
+@InternalApi private[akka] class RetryFlowCoordinator[In, CtxIn, Out, CtxOut](
+    minBackoff: FiniteDuration,
+    maxBackoff: FiniteDuration,
+    randomFactor: Double,
+    maxRetries: Int,
+    decideRetry: ((In, CtxIn), (Out, CtxOut)) => Option[(In, CtxIn)])
+    extends GraphStage[BidiShape[(In, CtxIn), (In, CtxIn), (Out, CtxOut), (Out, CtxOut)]] {
+
+  private val externalIn = Inlet[(In, CtxIn)]("RetryFlow.externalIn")
+  private val externalOut = Outlet[(Out, CtxOut)]("RetryFlow.externalOut")
+
+  private val internalOut = Outlet[(In, CtxIn)]("RetryFlow.internalOut")
+  private val internalIn = Inlet[(Out, CtxOut)]("RetryFlow.internalIn")
+
+  override val shape: BidiShape[(In, CtxIn), (In, CtxIn), (Out, CtxOut), (Out, CtxOut)] =
+    BidiShape(externalIn, internalOut, internalIn, externalOut)
+
+  override def createLogic(attributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
+
+    private var elementInProgress: Option[(In, CtxIn)] = None
+    private var retryNo = 0
+
+    setHandler(
+      externalIn,
+      new InHandler {
+        override def onPush(): Unit = {
+          val element = grab(externalIn)
+          elementInProgress = Some(element)
+          retryNo = 0
+          push(internalOut, element)
+        }
+
+        override def onUpstreamFinish(): Unit =
+          if (elementInProgress.isEmpty) {
+            completeStage()
+          }
+      })
+
+    setHandler(
+      internalOut,
+      new OutHandler {
+        override def onPull(): Unit = {
+          if (elementInProgress.isEmpty) {
```

Comment:
I think that if the internal flow has a buffer, and pulls here while there is still an element in flight, that pull is lost. See if you can cover that with a test case.